File: Arrays/Spiral Matrix.py
  """
  https://practice.geeksforgeeks.org/problems/spirally-traversing-a-matrix-1587115621/1#
  """
  
  def spirallyTraverse(self,matrix, r, c): 
        left = 0 #Indicates left boundary
        right = c-1 #Indicates right boundary
        top = 0 #Indicates top boundary
        bottom = r-1 #Indicates bottom boundary
        direction = 0
        result =[]
        while left<=right and top<=bottom:
            
            if direction == 0 :
                for i in range(left,right+1):
                    result.append(matrix[top][i])
            #Traversing the top boundary and incrementing top to point to new top boundary
                top+=1
                direction = 1
            
            elif direction == 1 :
                for i in range(top,bottom+1):
                    result.append(matrix[i][right])
            #Traversing the right boundary and decrementing right to point to new right boundary
                right-=1
                direction = 2
            
            elif direction == 2 :
                for i in range(right,left-1,-1):
                    result.append(matrix[bottom][i])
            #Traversing the bottom boundary and decrementing bottom to point to new bottom boundary
                bottom-=1
                direction = 3
            
            elif direction == 3 :
                for i in range(bottom,top-1,-1):
                    result.append(matrix[i][left])
                #Traversing the left boundary and incrementing left to point to new left boundary
                left+=1
                direction = 0
        
        return result
            

File: Arrays/Spiral Matrix.java
  """
  https://practice.geeksforgeeks.org/problems/spirally-traversing-a-matrix-1587115621/1#
  """
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws Exception {
        Scanner scn = new Scanner(System.in);
        int rows = scn.nextInt();
        int cols = scn.nextInt();
        int[][] matrix = new int[rows][cols];
        for(int i=0;i<rows;i++){
            for(int j=0;j<cols;j++){
                matrix[i][j] = scn.nextInt();
            }
        }
    //Spiral Code starts from here
        int rmin = 0;
        int rmax = rows-1;
        int cmin = 0;
        int cmax = cols-1;
        int tne = rows*cols;
        int count = 0;
        while(count<tne){
            for(int row=rmin;row<=rmax && count<tne;row++){
                System.out.println(matrix[row][cmin]);
                count++;
            }
            //First column is printed and therefore I can increment first column by 1.SImilarly for others
            cmin++;
            for(int col=cmin;col<=cmax && count<tne;col++){
                System.out.println(matrix[rmax][col]);
                count++;
            }
            rmax--;
            for(int row=rmax;row>=rmin && count<tne;row--){
                System.out.println(matrix[row][cmax]);
                count++;
            }
            cmax--;
            for(int col=cmax;col>=cmin && count<tne;col--){
                System.out.println(matrix[rmin][col]);
                count++;
            }
            rmin++;
            
        }
    }

}
